### PR TITLE
Add Device Plugin Integration to kubernetes 

### PR DESCRIPTION
**Pull request by RenaudWasTaken**
**What this PR does / why we need it**:
This implement the design discussed at https://github.com/kubernetes/community/pull/695

**Which issue this PR fixes** * https://github.com/kubernetes/community/pull/695

**Special notes for your reviewer**:
- It's probably easier to go through the commits rather than viewing all the files changed
  I've tried to logically separate them so that we can see what part of the code does what
- First proposal submitted to the community repo, please advise if something's not right with the format or procedure, etc.

@vishh @derekwaynecarr

**Release note**:
```release-note
External Hardware device discovery, monitoring and containerization 
```

### CODE REVIEW

**Comment by RenaudWasTaken** [comment]
Well that's a bummer all the tests failed :(

**Comment by deads2k** [comment]
@derekwaynecarr node related?

**Comment by RenaudWasTaken** [comment]
@deads2k it's more Resource Management workgroup related :)

**Comment by derekwaynecarr** [comment]
@deads2k -- we need agreement on the proposal before merging.  i will handle review for this.

**Comment by RenaudWasTaken** [comment]
In an effort to make the PR less important I removed:
- Protobuf generated files ~3500 LOC
- API changes (impacts the ability to rebuild state from a reboot) ~2500 LOC
- testing ~500 LOC

The PR is now "only" 1500 LOC

File: pkg/kubelet/cm/device_plugin_handler.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+	}
+
+	return nil
+}
+
+// Deallocate is the call that you can use to deallocate the devices assigned
+// to a container
+func (h *DevicePluginHandler) DeallocateDevices(p *v1.Pod, ctr string) {
+	devs := h.pod2Dev[p.UID][ctr]
+
```
Comment (lichuqiang) [inline pkg/kubelet/cm/device_plugin_handler.go]: Did not see deletion action in pod2Dev, is that needed?

File: pkg/kubelet/kuberuntime/kuberuntime_container.go [hunk 2]
```diff
@@ -594,6 +615,13 @@ func (m *kubeGenericRuntimeManager) killContainer(pod *v1.Pod, containerID kubec
 		glog.V(3).Infof("Container %q exited normally", containerID.String())
 	}
 
+	if m.containerManager != nil {
+		hdlr := m.containerManager.GetDevicePluginHandler()
+		if hdlr != nil {
+			hdlr.DeallocateDevices(pod, containerName)
```
Comment (lichuqiang) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: It seems that it does not call for func in client to release the device in the realization of DeallocateDevices， like it does in func AllocateDevices, can the devices get release gracefully?

**Comment by RenaudWasTaken** [comment]
Hi @lichuqiang !

Sorry I should have closed this PR earlier...
We decided to split it into multiple others. I'm looking into creating a tracking issue.
Right now there are two PRs:
- The Device Plugin Protobuf API: https://github.com/kubernetes/kubernetes/pull/49342
- The Device Plugin Protobuf scripts: https://github.com/kubernetes/kubernetes/pull/50401

**Pull request closed without merging**